File: algorithms/easy/count_symmetric_integers.ex
defmodule Solution do
  @spec count_symmetric_integers(low :: integer, high :: integer) :: integer
  def count_symmetric_integers(low, high) do
    low..high
    |> Enum.reduce(0, fn i, acc ->
      if is_symmetric(i) do
        acc + 1
      else
        acc
      end
    end)
  end

  def is_symmetric(n) do
    arr =
      "#{n}"
      |> String.graphemes()
      |> Enum.map(&String.to_integer/1)

    len = length(arr)

    if rem(len, 2) == 1 do
      false
    else
      {left, right} = arr |> Enum.split(div(len, 2))
      Enum.sum(left) == Enum.sum(right)
    end
  end
end

File: algorithms/easy/count_symmetric_integers.dart
bool isSymmetric(int n) {
  List<int> arr = [];
  while (n > 0) {
    arr.add(n % 10);
    n ~/= 10;
  }
  if (arr.length.isOdd) {
    return false;
  }

  int left = arr.sublist(0, arr.length ~/ 2).reduce((a, b) => a + b);
  int right = arr.sublist(arr.length ~/ 2).reduce((a, b) => a + b);
  return left == right;
}

class Solution {
  int countSymmetricIntegers(int low, int high) {
    int ct = 0;
    for (int i = low; i <= high; i++) {
      if (isSymmetric(i)) {
        ct++;
      }
    }
    return ct;
  }
}
